### PR TITLE
[FEATURE] annotate ion index and charge as subscript

### PR DESCRIPTION
**Pull request by timosachsenberg**
Added sub/superscripts to TOPPView Peakannotations.
Also added vertical + diagonal line drawing of peak annotations to (manually) reduce overlaps between annotations.

### CODE REVIEW

**Comment by timosachsenberg** [comment]
In TOPPView: prints ion index as subscript, charge as superscript
Common losses e.g. H2O are now also with 2 as subscript

**Comment by timosachsenberg** [comment]
@enetz and @hendrikweisser can you check if this change impacts visualization of your fragment ion annotations?

**Comment by timosachsenberg** [comment]
test

File: src/openms_gui/source/VISUAL/ANNOTATION/Annotation1DPeakItem.cpp [hunk 1]
```diff
@@ -196,20 +196,78 @@ namespace OpenMS
...
       delete(closest_ip);
     }
 
-    painter.drawText(bounding_box_, Qt::AlignCenter, text_);
+    // some pretty printing
+    QString text = text_;
+    if (!text.contains("<\\")) // don't process HTML strings again
+    {
+      // extract ion index
+      QRegExp reg_exp("[a|b|c|x|y|z](\\d+)");
```
Comment (enetz) [inline src/openms_gui/source/VISUAL/ANNOTATION/Annotation1DPeakItem.cpp]: according to the QRegExp documentation:
`[ABCD] will match an A or a B or a C or a D`
So you should not need the `|` bars, the `[]` brackets already imply `or` between the letters, I think.
Looks like you would also be matching strings like `|3+` right now.

File: src/openms_gui/source/VISUAL/ANNOTATION/Annotation1DPeakItem.cpp [hunk 2]
```diff
@@ -196,20 +196,78 @@ namespace OpenMS
...
+      text.replace(QRegExp("3\\+$"), "<sup>3+</sup>");
+      text.replace(QRegExp("\\+\\+$"), "<sup>2+</sup>");
+      text.replace(QRegExp("\\+$"), "");
+      text.replace(QRegExp("3\\-$"), "<sup>3-</sup>");
+      text.replace(QRegExp("\\-\\-$"), "<sup>2-</sup>");
+      text.replace(QRegExp("\\-$"), "");
+
+      // common losses
+      text.replace("H2O","H<sub>2</sub>O");
+      text.replace("H2O1","H<sub>2</sub>O");
```
Comment (enetz) [inline src/openms_gui/source/VISUAL/ANNOTATION/Annotation1DPeakItem.cpp]: The order is wrong, `H2O` is replaced first, and then `H<sub>2</sub>O1` will not be seen as `H2O1` and the `1` stays. This happens to my labels.
For `NH3` the two versions are sufficiently different, so this does not happen there.

Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/ANNOTATION/Annotation1DPeakItem.cpp]: aaah great catch! Thanks!

**Comment by enetz** [comment]
@timosachsenberg my labels are changed substantially by these changes, but since I like the idea and look in general, I adapted the code to work with my labels (PR to your branch).
Right now I have these issues:

- the colors of the labels were lost, my previously green and red labels are all black now. The peak colors are still there. I am actually not sure why this happened and whether this PR did this. The label color should not be affected by the changes here.
- I am not sure I agree, that just removing the single charge `+` and `-` is a good idea.

**Comment by timosachsenberg** [comment]
can you rebase your PR on my recent changes?

**Comment by timosachsenberg** [comment]
@enetz I tried to be closer to what is currently considered quasi-standard (in at least one subfield ;). 
See e.g., https://www.mcponline.org/content/early/2019/05/14/mcp.TIR118.001209 .
It is also the way they do it in Hennings lab.

**Comment by enetz** [comment]
@timosachsenberg I just noticed another case in the RNPxl test files:
e.g. `[M+H]-H2O++ U'+`, here the `++` is not at the end and isn't turned into a `2+` superscript.

Also in some cases, if the largest peak is labeled, the Y scale is normalized by that peak and the label is above the peak, outside of the spectrum. I noticed that by trying to remove all labels, can't see or select this one in some spectra.

I added a check for match position of the ion type again to exclude the cases without matches. It

**Comment by timosachsenberg** [comment]
Yes the old test files need an update. the string was not properly generated in the first place.

**Comment by enetz** [comment]
A similar issue with the automatically generated labels:
e.g. `b6+3\nYTPYLG`
the subscript for the ion index works, but the charge is at the end of the first line, not the whole string.
If we constrain the charge search to the first line, this would work.

**Comment by timosachsenberg** [comment]
Good point. So Eiter the first newline or end of string- whatever comes first. Can you add that?

**Comment by timosachsenberg** [comment]
k I think I fixed "Also in some cases, if the largest peak is labeled, the Y scale is normalized by that peak and the label is above the peak, outside of the spectrum. I noticed that by trying to remove all labels, can't see or select this one in some spectra."
as well as NASE annotations

**Comment by timosachsenberg** [comment]
let's merge and add the support for the automatically generated ones later

**Pull request merged**